### PR TITLE
DataProxy Service Definition

### PR DESCRIPTION
**Pull request by EngHabu**
Signed-off-by: Haytham Abuelfutuh <haytham@afutuh.com>

# TL;DR
Add dataproxy service definition and CreateUploadLocation API

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Tracking Issue

https://github.com/flyteorg/flyte/issues/2263

### CODE REVIEW

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]

**Comment by wild-endeavor** [comment]
the id... my only question is if there will ever in the future be a difference if i request `suffix=name` and `suffix=name/`.  I assume the idea is that for now, we'd request `suffix=fastabc123.tar.gz`?  Will it work if we request `suffix=aprefix/fastabc123.tar.gz`?

**Comment by EngHabu** [comment]
It works if you request `pre/fix/file.ext` yes... s3 and gcs (at least) only support a specific key though

File: protos/flyteidl/service/dataproxy.proto [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  // exceeds the platform allowed max.
+  google.protobuf.Duration expires_in = 4;
+}
+
+// DataProxy defines an RPC Service that allows access to user-data in a controlled manner.
+service DataProxy {
+  // Retrieves user information about the currently logged in user.
+  rpc CreateUploadLocation (CreateUploadLocationRequest) returns (CreateUploadLocationResponse) {
+    option (google.api.http) = {
+      post: "/api/v1/dataproxy/upload"
```
Comment (kumare3) [inline protos/flyteidl/service/dataproxy.proto:43]: I dont like this api, it has a verb at the end `upload`. I would prefer a noun. For example even the "upload_location" is a noun.
How about, artifact_urn? or artifact_loc

File: protos/flyteidl/service/dataproxy.proto [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  // Domain to create the upload location for
+  string domain = 2;
+
+  // Suffix specifies a desired suffix for the generated location. E.g. `/file.py` or `pre/fix/file.zip`.
+  string suffix = 3;
+
+  // ExpiresIn defines a requested expiration duration for the generated url. The request will be rejected if this
+  // exceeds the platform allowed max.
+  google.protobuf.Duration expires_in = 4;
```
Comment (kumare3) [inline protos/flyteidl/service/dataproxy.proto:35]: this should be optional right? can we label it so?

File: protos/flyteidl/service/dataproxy.proto [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+// CreateUploadLocationRequest specified request for the CreateUploadLocation API.
+message CreateUploadLocationRequest {
+  // Project to create the upload location for
+  string project = 1;
+
+  // Domain to create the upload location for
+  string domain = 2;
+
+  // Suffix specifies a desired suffix for the generated location. E.g. `/file.py` or `pre/fix/file.zip`.
+  string suffix = 3;
```
Comment (kumare3) [inline protos/flyteidl/service/dataproxy.proto:31]: can we make this optional and if so label it?

**Review by wild-endeavor — APPROVED** [review]

**Review by kumare3 — APPROVED** [review]

**Pull request merged**